content/r-package-dev/Writing_Tests.Rmd

```diff
@@ -0,0 +1,153 @@
+
+**Behavior driven development (BDD):** Functionality is declared in clear language, specifying the requirements of the functionality.  Tests are then written, tied to this language, to verify that this behavior is met.
+
+**Unit testing:** Testing the individual units of functionality isolated from the system to ensure that those units perform the required functionality correctly.  This level of testing forms the safety net that ensures new code can be added with little risk of effecting other parts of the system, or old code can be modified while maintaining its intended function.
+
+**Integration testing:** Testing several units in combination to make sure they integrate properly.  This involves testing higher level functionality that touches several components and ensuring that it is correct.  This testing often takes more setup and time to run, but is valuable to ensure proper testing of the system as a whole.
+
+**Characterization testing:** Testing the behavior of code as it exists to ensure that it continues to behave in that way.  This technique is often used on older code where the expected functionality is not fully clear.  Write tests that exercise the functions as they exist and use the result as the expectation.  Changes can then be made to the code without changing the behavior, and new tests can be written to better explain the functions.
+
+**Data testing:** Data is often treated differently than code, but it can be tested in quite similar ways.  Assumptions about data types, valid ranges, and other aspects of the data should be explicitly tested for to ensure validity.
+
+Applying these techniques where applicable will develop a solid suite of tests that will instill confidence in those developing on or against your package.  It is helpful to gather code coverage metrics to quantify the level of testing that exists.  The R package `covr` is available to generate these metrics and push them to a location to share the results.
+
+## How do I write tests?
+
+We are going to build on what was covered in the documentation section by writing some tests.  Here is a reminder of the pH validity function we created earlier.
+
+```{r ph-setup, include=TRUE, warning=FALSE, message=FALSE}
+is.valid.pH <- function(pH){
+  pH >= 0 & pH <= 14
+}
+```
+
+```{r reporter-setup, include=FALSE, warning=FALSE, message=FALSE}
+  reporter <- testthat:::find_reporter("list")
+  old <- testthat:::set_reporter(reporter)
+```
+
+The `testthat` package provides the functions we are going to use for testing.  Files performing the testing should be placed in the `tests/testthat` directory and a `testhat.R` script should invoke `test_check()` on your package.
```

Comment:
also consider saying something about helper files that provide shared code across tests? i use them a lot, but not sure whether they'd be a key detail or a distraction for a new student/user of testthat...